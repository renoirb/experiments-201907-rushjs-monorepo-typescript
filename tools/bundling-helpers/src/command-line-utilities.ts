
import { log } from './utils'

const hasCliArg = (name: string): boolean => {
  const context: NodeJS.Process = process || {}
  const { argv = [] } = context
  return argv.includes(name)
}

/**
 * @param {string} name — NodeJS.Process CLI argv name
 */
const getCliArg = (name: string): false | [string, string | true] => {
  if (hasCliArg(name)) {
    const context: NodeJS.Process = process || {}
    const { argv = [] } = context
    const argIndex = [...argv].indexOf(name)
    const argValIndex = argIndex + 1
    const argName = Reflect.has(argv, argIndex)
      ? Reflect.get(argv, argIndex)
      : false
    let argVal = Reflect.has(argv, argValIndex)
      ? Reflect.get(argv, argValIndex)
      : false
    // If `--help --foo`, we do not want the next argument
    if (/^--/.test(argVal)) {
      argVal = true
    }
    if (argName && argVal) {
      return [argName, argVal]
    }
  }

  return false
}

/**
 * Check if process.env has a flag we are looking for
 *
 * https://docs.gitlab.com/ee/ci/variables/#syntax-of-environment-variables-in-job-scripts
 *
 * @param {string} name — name of the environment variable
 */
const hasEnvFlag = (name: string): boolean => {
  const context: NodeJS.Process = process || {}
  const { env = {} } = context
  const processEnvKeys = Object.keys(env)
  return processEnvKeys.includes(name)
}

/**
 * Command Line (CLI) Utilities.
 *
 * Helpers so we can avoid boilerplate CLI checks
 */
export class CommandLineUtilities {
  static getCliArg = getCliArg
  static hasCliArg = hasCliArg
  static hasEnvFlag = hasEnvFlag
  static log = log
}
