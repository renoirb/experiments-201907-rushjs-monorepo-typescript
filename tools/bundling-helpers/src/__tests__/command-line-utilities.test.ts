/* eslint-env jest */

import { CommandLineUtilities } from '../command-line-utilities'

describe('getProcessCliArgValuePair', () => {
  let oldArgv: string[] = []

  beforeAll(() => {
    oldArgv = [...process.argv]
    process.argv = [
      '--foo',
      'bar',
      '--buz',
      '--config',
      '/some/ficticious/path/to/file.json',
    ]
  })

  afterAll(() => {
    process.argv = oldArgv
  })

  it('Snapshot', () => {
    const subject = CommandLineUtilities.getCliArg('--config')
    expect(subject).toMatchObject([
      '--config',
      '/some/ficticious/path/to/file.json',
    ])
    expect(subject).toMatchSnapshot()
  })

  it('Missing CLI arguments returns false', () => {
    const subject = CommandLineUtilities.getCliArg('--not-present-arg')
    expect(subject).toBe(false)
  })

  it('Empty CLI arguments returns false', () => {
    const subject = CommandLineUtilities.getCliArg(' ')
    expect(subject).toBe(false)
  })

  it(`Argument less CLI argument if present is tuple name and value as boolean true ['--buz', true]`, () => {
    const subject = CommandLineUtilities.getCliArg('--buz')
    expect(subject).toMatchObject(['--buz', true])
  })

  it(`CLI argument with non value returns tuple name and value`, () => {
    const subject = CommandLineUtilities.getCliArg('--foo')
    expect(subject).toMatchObject(['--foo', 'bar'])
  })
})
