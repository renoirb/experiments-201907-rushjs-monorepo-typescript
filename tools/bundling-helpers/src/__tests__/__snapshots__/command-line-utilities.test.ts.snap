// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProcessCliArgValuePair Snapshot 1`] = `
Array [
  "--config",
  "/some/ficticious/path/to/file.json",
]
`;
